File: netlify/functions/feed.js
/**
 * Netlify JavaScript serverless function to fetch and parse the Substack RSS feed.
 * Deployed at /.netlify/functions/feed  →  proxied to /api/feed via netlify.toml
 */

const FEEDS = {
  substack: 'https://manikhanuja.substack.com/feed',
};

const CORS_HEADERS = {
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
};

function respond(statusCode, body) {
  return { statusCode, headers: CORS_HEADERS, body: JSON.stringify(body) };
}

function stripHtml(html) {
  return (html || '').replace(/<[^>]+>/g, '').replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&nbsp;/g, ' ').trim();
}

function getTag(xml, tag, after = 0) {
  // Grab the first occurrence of <tag>...</tag> after position `after`
  const open = xml.indexOf(`<${tag}`, after);
  if (open === -1) return '';
  const close = xml.indexOf(`</${tag}>`, open);
  if (close === -1) return '';
  const inner = xml.slice(xml.indexOf('>', open) + 1, close);
  return inner.replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, '$1').trim();
}

function parseRss(xml) {
  const items = [];
  let pos = 0;

  while (items.length < 3) {
    const start = xml.indexOf('<item>', pos);
    if (start === -1) break;
    const end = xml.indexOf('</item>', start);
    if (end === -1) break;
    const chunk = xml.slice(start, end + 7);
    pos = end + 7;

    const title   = stripHtml(getTag(chunk, 'title'));
    const link    = getTag(chunk, 'link') || getTag(chunk, 'guid');
    const pubDate = getTag(chunk, 'pubDate');
    const desc    = stripHtml(
      getTag(chunk, 'content:encoded') || getTag(chunk, 'description')
    );
    const excerpt = desc.length > 220 ? desc.slice(0, 220) + '…' : desc;

    let year = '2026';
    try { year = String(new Date(pubDate).getFullYear()); } catch (_) {}

    items.push({ title, link, year, excerpt });
  }

  return items;
}

exports.handler = async (event) => {
  // Handle CORS preflight
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 204, headers: CORS_HEADERS, body: '' };
  }

  try {
    const source = (event.queryStringParameters || {}).source || 'substack';
    const url = FEEDS[source];

    if (!url) {
      return respond(404, { ok: false, error: `Unknown source: ${source}` });
    }

    const res = await fetch(url, {
      headers: { 'User-Agent': 'Mozilla/5.0 (compatible; netlify-function/1.0)' },
      signal: AbortSignal.timeout(10000),
    });

    if (!res.ok) {
      throw new Error(`Upstream HTTP ${res.status}`);
    }

    const xml = await res.text();
    const items = parseRss(xml);

    if (!items.length) {
      throw new Error('No items found in feed');
    }

    return respond(200, { ok: true, items });

  } catch (err) {
    console.error('feed function error:', err.message);
    return respond(500, { ok: false, error: err.message });
  }
};

File: netlify/functions/feed.py
"""
Netlify Python serverless function to fetch and parse RSS feeds.
Handles CORS and returns JSON format suitable for the frontend.
"""

import json
import urllib.request
import xml.etree.ElementTree as ET
from datetime import datetime

FEEDS = {
    'substack': 'https://manikhanuja.substack.com/feed',
}

NS = {
    'content': 'http://purl.org/rss/1.0/modules/content/',
    'dc': 'http://purl.org/dc/elements/1.1/',
    'atom': 'http://www.w3.org/2005/Atom',
}


def fetch_rss(url):
    """Fetch RSS feed from URL with proper headers."""
    req = urllib.request.Request(
        url,
        headers={'User-Agent': 'Mozilla/5.0 (compatible; netlify-function/1.0)'}
    )
    with urllib.request.urlopen(req, timeout=10) as resp:
        return resp.read()


def parse_rss(xml_bytes):
    """Parse RSS XML and extract items."""
    root = ET.fromstring(xml_bytes)
    channel = root.find('channel')
    items = []

    for item in (channel.findall('item') if channel is not None else root.findall('.//item'))[:3]:
        def t(tag, ns=None):
            el = item.find(f'{{{ns}}}{tag}' if ns else tag)
            return (el.text or '').strip() if el is not None else ''

        desc = t('encoded', NS['content']) or t('description')

        # Strip HTML tags
        import re
        clean = re.sub(r'<[^>]+>', '', desc).strip()
        excerpt = clean[:200] + ('...' if len(clean) > 200 else '')

        pub = t('pubDate')
        try:
            year = str(datetime.strptime(pub, '%a, %d %b %Y %H:%M:%S %z').year)
        except Exception:
            year = '2026'

        items.append({
            'title': t('title'),
            'link': t('link'),
            'year': year,
            'excerpt': excerpt,
        })

    return items


def handler(event, context):
    """Netlify function handler for RSS feed endpoint."""
    try:
        # Get source from query parameters
        query_params = event.get('queryStringParameters', {}) or {}
        source = query_params.get('source', 'substack')

        # Get the feed URL
        url = FEEDS.get(source)
        if not url:
            return {
                'statusCode': 404,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                },
                'body': json.dumps({'ok': False, 'error': f'Unknown source: {source}'})
            }

        # Fetch and parse the feed
        xml_bytes = fetch_rss(url)
        items = parse_rss(xml_bytes)

        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'ok': True, 'items': items})
        }

    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'ok': False, 'error': str(e)})
        }
